src/core/types.ts

```diff
@@ -62,6 +62,11 @@ export interface QueryOptions<
   initialData?: TData | InitialDataFunction<TData>
   initialDataUpdatedAt?: number | (() => number | undefined)
   behavior?: QueryBehavior<TQueryFnData, TError, TData>
+  /**
+   * This callback will fire right before the query will start fetching new data.
+   * Useful for side-effects that are specific to this query.
+   */
+  onQuery?: (context:QueryFunctionContext) => void
```

Comment:
should this not be defined on the `QueryObserverOptions`, like the other callbacks (`onSuccess`, `onError`, `onSettled`) ? We want to be able to have different callbacks per observer

Reply:
Thank you! These are valid points.

- I moved `onQuery` to `QueryObserverOptions`.
- When `onQuery` exists,  we're calling it from `queryFn()` now, instead of the retryer.
    - It makes the code a lot simpler: no need to modify infiniteQueryBehavior anymore.